Improve compression of Arrays of Integers (High compression mode)

Cyan4973 opened the pull request:
This is an answer to issues similar to #3793 and #3014,
which report a poor compression ratio at high compression level
on synthetic data, typically arrays of 32-bit integers with very little consecutive variation (like an increasing series).
While such synthetic scenario tends to be "too regular", and therefore not representative of less regular "real" data,
it's nonetheless frequently used as a test scenario, because it is easy to produce.

The High compression mode has indeed a weakness in such a scenario,
and this patch tries to address it.
A first patch was able to improve these synthetic scenarios, increasing compression ratio by a significant amount. But unfortunately, it would then lose some compression ratio for other test data.
This patch is a variant, which manages to improve compression ratio in the case of 32-bit integer arrays, without losing anything in other cases.

This is illustrated in below test cases : 
| file | raw | `dev` | this `PR` | Improvement |
| --- | ---:| ---:| ---:| --- |
| inc_1M | 4000000 | 2687382 | 831424 | x3.23 | 
| issue3793 | 262144 | 142849 | 82866 | x1.72 |
| `.debug_str_offsets` | 296304 | 179814 | 131410 | +36.8% |
| canterbury/kennedy.xls | 1029744 | 69724 | 64825 | +7.5% |
| calgary/geo | 102400 | 64712 | 63051 | +2.6% |
| silesia/x-ray | 8474240 | 5155749 | 5129823 | +0.5% | 
| silesia/samba | 21606400 |  3907283 | 3899439 | +0.2% |
| silesia.tar | 211957760 | 52988966 | 52877671 | +0.2%  |

As can be seen, compression benefits are very sensible on the synthetic examples. More realistic samples (such as standard compression corpuses) get lower yet noticeable gains, notably `.debug_str_offsets`, `kennedy.xls`, `geo` and `x-ray`. Several other files get smaller gains (+0.1% range), but more importantly, no file in any compression corpus is negatively impacted by this patch (beyond a few bytes here and there just as a consequence of random luck). 
Overall, the resulting outcome of this patch on `silesia` at level 19 is > 100 KB savings. 

As a side note, I was hoping for an even better outcome because the gains from an earlier patch for 32-bit integer arrays were larger. But this previous patch was also negatively impacting compression ratio of other (non 32-bit integers) files. Therefore I believe this positive-only patch is a more acceptable trade-off for merge.

I might try later to reproduce the stronger results of the first patch (on 32-bit integer arrays), with the hope to find a solution which doesn't negatively impact other data. But this may require a complete overhaul of the optimal parser, and in the meantime, we can  at least use this patch.

Cyan4973 committed:
improve high compression ratio for file like #3793

this works great for 32-bit arrays,
notably the synthetic ones, with extreme regularity,
unfortunately, it's not universal,
and in some cases, it's a loss.
Crucially, on average, it's a loss on silesia.
The most negatively impacted file is x-ray.
It deserves an investigation before suggesting it as an evolution.

Cyan4973 committed:
refactor optimal parser

store stretches as intermediate solution instead of sequences.
makes it possible to link a solution to a predecessor.

Cyan4973 committed:
modification: differentiate literal update at pos+1

helps when litlen==1 is cheaper than litlen==0

works great on pathological arr[u32] examples
but doesn't generalize well on other files.

silesia/x-ray is amoung the most negatively affected ones.

Cyan4973 committed:
finally, a version that generalizes well

While it's not always strictly a win,
it's a win for files that see a noticeably compression ratio increase,
while it's a very small noise for other files.

Downside is, this patch is less efficient for 32-bit arrays of integer
than the previous patch which was introducing losses for other files,
but it's still a net improvement on this scenario.

Cyan4973 committed:
minor optimization, mostly for clarity

Cyan4973 committed:
fixed wrong assert

Cyan4973 committed:
update compression results

good news: there are only improvements

Cyan4973 committed:
fixed wrong assert

by introducing ZSTD_OPT_SIZE

Cyan4973 committed:
removed trace control

Cyan4973 committed:
use ZSTD_memcpy()

which can be redirected in Linux kernel mode

Cyan4973 commented:
So there is still a reported problem ("use-after-poison") in one of the tests (`asan-ubsan-regression`), 
but, while I can reproduce the error locally on a linux workstation,
I'm currently unable to extract any trace from it,
which is hampering deeper investigation.

_edit_ : now fixed, but could probably have been easier.

Cyan4973 committed:
fix uasan dictionary_stream_round_trip fuzz test

Cyan4973 committed:
fix msan warnings

Cyan4973 committed:
update compression results for regression tests

terrelln commented on lib/compress/zstd_opt.c:
```diff
@@ -1143,38 +1148,41 @@ ZSTD_compressBlock_opt_generic(ZSTD_matchState_t* ms,
-                    lastSequence.mlen = maxML;
-                    lastSequence.off = maxOffBase;
-                    DEBUGLOG(6, "large match (%u>%u), immediate encoding",
+                    lastStretch.litlen = 0;
```
I don't understand this change.

Edit: After reading the code I understand. It would've been helpful to see a description of the change that this diff made in the summary, because it was non-obvious until I finished reading the code.

terrelln commented on lib/compress/zstd_opt.c:
```diff
@@ -1271,63 +1297,104 @@ ZSTD_compressBlock_opt_generic(ZSTD_matchState_t* ms,
+         * - So far, @opt stored stretches, aka a match followed by literals
+         * - Now, it will store sequences, aka literals followed by a match
+         */
+        {   U32 const storeEnd = cur + 2;
```
Why +2? Also update the docs which say that `storeEnd == cur+1`.

Cyan4973 replied:
Note below, if `lastStretch.litlen > 0`, i.e. there are some literals _after_ the last match,
then we start to write into `opt[storeStart]` with `storeStart == storeEnd-1`,
therefore overwriting the content of `opt[storeStart]`. And if `storeEnd==cur+1`, then `storeStart` would be `==cur`.
However, `opt[cur]` is where the next _stretch_ is stored, and we haven't read it yet, since we are still dealing with `lastStretch`. So we would be overwriting a component of our next sequence.

A `+2` avoids this overwriting.
This is fine because `ZSTD_OPT_SIZE == ZSTD_OPT_NUM + 2`, so we have enough margin, even when the array is entirely filled. This is checked in the following `assert` : `assert(storeEnd < ZSTD_OPT_SIZE);`.

terrelln commented on lib/compress/zstd_opt.c:
```diff
@@ -1185,40 +1193,60 @@ ZSTD_compressBlock_opt_generic(ZSTD_matchState_t* ms,
+                    opt[cur] = opt[cur-1];
                     opt[cur].litlen = litlen;
                     opt[cur].price = price;
+                    if ( (optLevel >= 1) /* additional check only for higher modes */
+                      && (prevMatch.litlen == 0) /* replace a match */
+                      && (LL_INCPRICE(1) < 0) /* ll1 is cheaper than ll0 */
+                    ) {
+                        /* check next position, in case it would be cheaper */
+                        int with1literal = prevMatch.price + LIT_PRICE(ip+cur) + LL_INCPRICE(1);
+                        int withMoreLiterals = price + LIT_PRICE(ip+cur) + LL_INCPRICE(litlen+1);
+                        DEBUGLOG(7, "then at next rPos %u : match+1lit %.2f vs %ulits %.2f",
+                                cur+1, ZSTD_fCost(with1literal), litlen+1, ZSTD_fCost(withMoreLiterals));
+                        if ( (with1literal < withMoreLiterals)
+                          && (with1literal < opt[cur+1].price) ) {
+                            /* update offset history - before it disappears */
+                            U32 const prev = cur - prevMatch.mlen;
+                            repcodes_t const newReps = ZSTD_newRep(opt[prev].rep, prevMatch.off, opt[prev].litlen==0);
+                            assert(cur >= prevMatch.mlen);
+                            DEBUGLOG(7, "==> match+1lit is cheaper (%.2f < %.2f) (hist:%u,%u,%u) !",
+                                        ZSTD_fCost(with1literal), ZSTD_fCost(withMoreLiterals),
+                                        newReps.rep[0], newReps.rep[1], newReps.rep[2] );
+                            opt[cur+1] = prevMatch;  /* mlen & offbase */
+                            ZSTD_memcpy(opt[cur+1].rep, &newReps, sizeof(repcodes_t));
+                            opt[cur+1].litlen = 1;
+                            opt[cur+1].price = with1literal;
+                            if (last_pos < cur+1) last_pos = cur+1;
+                        }
+                    }
```
Besides the change to `litlength`, which is to facilitate this change, this section is the major functional change of this diff, correct?

Cyan4973 replied:
> Besides the change to `litlength`, which is to facilitate this change, this section is the major functional change of this diff, correct?

Yes

Cyan4973 committed:
added or updated code comments

as suggested by @terrelln,
to make the code of the optimal parser a bit more understandable.

ribasushi commented:
Hi @Cyan4973, excellent work! Could you please clarify which "High compression mode" levels benefit from this work? Only `19`? `17+`? `15+`? 

I tried to figure this out from the changelog/code diff, but either I am missing it or it is just unclear...

Cyan4973 commented:
It think it's active for strategies `bt_ultra` and `bt_ultra2`.
Depending on the size of the input, these strategies are triggered at different compression levels.
But assuming the file is "large", this should be levels 18 and above.
